---
title: 'Omic Characterization of Doxorubicin Resistant Bone Sarcoma Models: PhosphoProteomic'
author: "Borja Gallego Mart√≠nez"
date: "`r Sys.Date()`"
output: 
  html_document: 
    toc: true
    toc_float: true
    toc_collapsed: true
    highlight: pygments
    theme: paper
    df_print: kable
---

```{r Setup, include=FALSE}
 # Setup default parameters in the code chunks
knitr::opts_chunk$set(
  echo = F, 
  message = F, 
  warning = F, 
  eval = T, 
  cache.lazy = F
  )
```

```{r Packages and Functions Setup, message=FALSE, warning=FALSE, include=FALSE}
library(tidyverse)
library(readxl)
library(plotly)
library(limma)
#library(PhosR)
library(ggrepel)
library(ggVennDiagram)
library(VennDiagram)
library(OmnipathR)
library(decoupleR)
select <- dplyr::select
mutate <- dplyr::mutate
source("00_Functions.R")
set.seed(123)
```

# **Importing Input Data**

```{r Metadata table}
Metadata <- read_xlsx("./Raw/Phospho/Phos_Metadata.xlsx") %>%
  column_to_rownames(var = "Label")
Metadata
```

```{r Proccessing Input Data}
b143.raw <- read_xlsx("./Raw/Phospho/MS_results_PRC-5343 143B.xlsx", sheet = 3,
                      range = cell_cols("G:AM")) %>%
  filter(is.na(`Potential contaminant`)) %>%
  mutate(
    Symbol = sapply(strsplit(`Gene names`, ";"), "[[", 1)
  ) %>%
  column_to_rownames(var = "PTMpeptide_IDs")
b143.counts <- b143.raw[, grep("log2", colnames(b143.raw))]
colnames(b143.counts) <- rownames(Metadata[Metadata$cell_line == "b143", ])
#b143.ppe <- PhosphoExperiment(
#  assays = list(Quantification = as.matrix(b143.counts)),
#  colData = Metadata[Metadata$cell_line == "b143", ],
#  Site = b143.raw$Position,
#  GeneSymbol = b143.raw$Symbol,
#  Residue = b143.raw$`Amino acid`,
#  Sequence = sapply(strsplit(as.character(b143.raw[, "Sequence window"]), ";"),
#                    function(x){x[1]}),
#  UniprotID = sapply(strsplit(b143.raw$`Leading proteins`, ";"), "[[", 1),
#  Localisation = b143.raw$`Localization prob`
#)
Saos2.raw <- read_xlsx("./Raw/Phospho/MS_results_PRC-5343 Saos2.xlsx", sheet = 3,
                       range = cell_cols("G:AM")) %>%
  filter(is.na(`Potential contaminant`)) %>%
  mutate(
    Symbol = sapply(strsplit(`Gene names`, ";"), "[[", 1)
  ) %>%
  column_to_rownames(var = "PTMpeptide_IDs")
Saos2.counts <- Saos2.raw[, grep("log2", colnames(Saos2.raw))]
colnames(Saos2.counts) <- rownames(Metadata[Metadata$cell_line == "Saos2", ])
#Saos2.ppe <- PhosphoExperiment(
#  assays = list(Quantification = as.matrix(Saos2.counts)),
#  colData = Metadata[Metadata$cell_line == "Saos2", ],
#  Site = Saos2.raw$Position,
#  GeneSymbol = Saos2.raw$Symbol,
#  Residue = Saos2.raw$`Amino acid`,
#  Sequence = sapply(strsplit(as.character(Saos2.raw[, "Sequence window"]), ";"),
#                    function(x){x[1]}),
#  UniprotID = sapply(strsplit(Saos2.raw$`Leading proteins`, ";"), "[[", 1),
#  Localisation = Saos2.raw$`Localization prob`
#)
TCDS17.raw <- read_xlsx("./Raw/Phospho/MS_results_PRC-5343 TCDS17.xlsx", sheet = 3,
                        range = cell_cols("G:AM")) %>%
  filter(is.na(`Potential contaminant`)) %>%
  mutate(
    Symbol = sapply(strsplit(`Gene names`, ";"), "[[", 1)
  ) %>%
  column_to_rownames(var = "PTMpeptide_IDs")
TCDS17.counts <- TCDS17.raw[, grep("log2", colnames(TCDS17.raw))]
colnames(TCDS17.counts) <- rownames(Metadata[Metadata$cell_line == "TCDS17", ])
#TCDS17.ppe <- PhosphoExperiment(
#  assays = list(Quantification = as.matrix(TCDS17.counts)),
#  colData = Metadata[Metadata$cell_line == "TCDS17", ],
#  Site = TCDS17.raw$Position,
#  GeneSymbol = TCDS17.raw$Symbol,
#  Residue = TCDS17.raw$`Amino acid`,
#  Sequence = sapply(strsplit(as.character(TCDS17.raw[, "Sequence window"]), ";"),
#                    function(x){x[1]}),
#  UniprotID = sapply(strsplit(TCDS17.raw$`Leading proteins`, ";"), "[[", 1),
#  Localisation = TCDS17.raw$`Localization prob`
#)
```

# **Quality Control**

## Correlation Heatmaps

```{r Correlation Heatmaps, message=FALSE}
b143.QC.CorrHeatmap <- QC.MakeCorrHeatmap(
  Counts.norm = b143.counts,
  heatmap.rows = colnames(b143.counts),
  main = "Correlation Heatmap of 143B Samples",
  output.path = "./Results/Phospho/Figures/b143_QC_CorrHeatmap.tiff"
    )
Saos2.QC.CorrHeatmap <- QC.MakeCorrHeatmap(
  Counts.norm = Saos2.counts,
  heatmap.rows = colnames(Saos2.counts),
  main = "Correlation Heatmap of SaOS-2 Samples",
  output.path = "./Results/Phospho/Figures/Saos2_QC_CorrHeatmap.tiff"
    )
TCDS17.QC.CorrHeatmap <- QC.MakeCorrHeatmap(
  Counts.norm = TCDS17.counts,
  heatmap.rows = colnames(TCDS17.counts),
  main = "Correlation Heatmap of T-CDS-17#4 Samples",
  output.path = "./Results/Phospho/Figures/TCDS17_QC_CorrHeatmap.tiff"
    )
```

## Principal Component Analysis

```{r PCAs, message=FALSE}
b143.QC.PCA <- QC.MakePCA(
  data = b143.counts,
  metadata = Metadata[Metadata$cell_line == "b143", ],
  main = "PCA of 143B Samples",
  output.path = "./Results/Phospho/Figures/b143_QC_PCA.tiff",
  fill.color = c("#FF0000", "#0000FF")
)
ggplotly(b143.QC.PCA$Plot)
Saos2.QC.PCA <- QC.MakePCA(
  data = Saos2.counts,
  metadata = Metadata[Metadata$cell_line == "Saos2", ],
  main = "PCA of SaOS-2 Samples",
  output.path = "./Results/Phospho/Figures/Saos2_QC_PCA.tiff",
  fill.color = c("#FF0000", "#0000FF")
)
ggplotly(Saos2.QC.PCA$Plot)
TCDS17.QC.PCA <- QC.MakePCA(
  data = TCDS17.counts,
  metadata = Metadata[Metadata$cell_line == "TCDS17", ],
  main = "PCA of T-CDS-17#4 Samples",
  output.path = "./Results/Phospho/Figures/TCDS17_QC_PCA.tiff",
  fill.color = c("#FF0000", "#0000FF")
)
ggplotly(TCDS17.QC.PCA$Plot)
```

# **Differential Expression Analysis**

```{r DEA: Tables}
b143.DEPhos <- DEA.limma2Table(
  data = b143.counts,
  condition = Metadata[Metadata$cell_line == "b143", "condition"],
  cont = c("DXR-Par"),
  output.csv = "./Results/Phospho/b143_DEPhosTable.csv",
  output.xlsx = "./Results/Phospho/b143_DEPhosTable.xlsx"
)
Saos2.DEPhos <- DEA.limma2Table(
  data = Saos2.counts,
  condition = Metadata[Metadata$cell_line == "Saos2", "condition"],
  cont = c("DXR-Par"),
  output.csv = "./Results/Phospho/Saos2_DEPhosTable.csv",
  output.xlsx = "./Results/Phospho/Saos2_DEPhosTable.xlsx"
)
TCDS17.DEPhos <- DEA.limma2Table(
  data = TCDS17.counts,
  condition = Metadata[Metadata$cell_line == "TCDS17", "condition"],
  cont = c("DXR-Par"),
  output.csv = "./Results/Phospho/TCDS17_DEPhosTable.csv",
  output.xlsx = "./Results/Phospho/TCDS17_DEPhosTable.xlsx"
)
```

## Volcano Plots

```{r DEA: Volcano Plots, message=FALSE}
b143.DEA.volcano <- DEA.MakeVolcano(
  DEA.table = b143.DEPhos,
  sym.coln = "row", lfc.coln = "logFC", fdr.coln = "adj.P.Val", lfc.thr = .5,
  main = "DEPhos for 143B DX-R vs. Parental", top = 0,
  output.path = "./Results/Phospho/Figures/b143_DEPhos_volcano.tiff"
)
b143.DEA.volcano
Saos2.DEA.volcano <- DEA.MakeVolcano(
  DEA.table = Saos2.DEPhos,
  sym.coln = "row", lfc.coln = "logFC", fdr.coln = "adj.P.Val", lfc.thr = .5,
  main = "DEPhos for SaOS-2 DX-R vs. Parental", top = 0,
  output.path = "./Results/Phospho/Figures/Saos2_DEPhos_volcano.tiff"
)
Saos2.DEA.volcano
TCDS17.DEA.volcano <- DEA.MakeVolcano(
  DEA.table = TCDS17.DEPhos,
  sym.coln = "row", lfc.coln = "logFC", fdr.coln = "adj.P.Val", lfc.thr = .5,
  main = "DEPhos for T-CDS-17#4 DX-R vs. Parental", top = 0,
  output.path = "./Results/Phospho/Figures/TCDS17_DEPhos_volcano.tiff"
)
TCDS17.DEA.volcano
```

## Checking Coincidences by Venn Diagrams

```{r DEA: Venn Diagrams, message=FALSE, warning=FALSE}
Overall.DEA.VennDiagrams <- DEA.MakeVenn(
  data.l = list(
    "143B" = b143.DEPhos,
    "SaOS-2" = Saos2.DEPhos,
    "T-CDS-17#4" = TCDS17.DEPhos
    ),
  lfc.coln = "logFC", fdr.coln = "adj.P.Val", Symbol = "row", lfc.thr = .5,
  up.main = "Up-Regulated Coincidences",
  up.file = "./Results/Phospho/Figures/Overall_DEA_VennDiagram_up.tiff",
  down.main = "Down-Regulated Coincidences",
  down.file = "./Results/Phospho/Figures/Overall_DEA_VennDiagram_down.tiff"
  )
Overall.DEA.VennDiagrams$Up.plot
data.frame("Global Up Coincidences" = Overall.DEA.VennDiagrams$Up.data[[1]])
Overall.DEA.VennDiagrams$Down.plot
data.frame("Global Down Coincidences" = Overall.DEA.VennDiagrams$Down.data[[1]])
```

# **Functional Analysis**

## Kinases Activity Inference

```{r Functional Analysis: Kinase activity inference, message=FALSE, warning=FALSE}
Kin.net <- MakeKinNet()
b143.Kin <- Get_Kinase(mat = b143.DEPhos, net = Kin.net, pval.thr = .05,
                       main = "Significant Kinases Regulated in 143B DX-R",
                       barplot.o = "./Results/Phospho/Figures/b143_Kinases.tiff",
                       method = "ULM", condition = "t")
ggplotly(b143.Kin$Kinases.barplot)
Saos2.Kin <- Get_Kinase(mat = Saos2.DEPhos, net = Kin.net, pval.thr = .05,
                        main = "Significant Kinases Regulated in SaOS-2 DX-R",
                        barplot.o = "./Results/Phospho/Figures/Saos2_Kinases.tiff",
                        method = "ULM", condition = "t")
ggplotly(Saos2.Kin$Kinases.barplot)
TCDS17.Kin <- Get_Kinase(mat = TCDS17.DEPhos, net = Kin.net, pval.thr = .05,
                         main = "Significant Kinases Regulated in T-CDS-17#4 DX-R",
                         barplot.o = "./Results/Phospho/Figures/TCDS17_Kinases.tiff",
                         method = "ULM", condition = "t")
ggplotly(TCDS17.Kin$Kinases.barplot)


b143.Kin_all <- Get_Kinase(
  mat = b143.DEPhos, net = Kin.net,
  main = "Kinases Regulated in 143B DX-R",
  barplot.o = "./Results/Phospho/Figures/b143_Kinases_all.tiff",
  method = "ULM", condition = "t"
  )
Saos2.Kin_all <- Get_Kinase(
  mat = Saos2.DEPhos, net = Kin.net,
  main = "Kinases Regulated in SaOS-2 DX-R",
  barplot.o = "./Results/Phospho/Figures/Saos2_Kinases_all.tiff",
  method = "ULM", condition = "t"
  )
TCDS17.Kin_all <- Get_Kinase(
  mat = TCDS17.DEPhos, net = Kin.net,
  main = "Kinases Regulated in T-CDS-17#4 DX-R",
  barplot.o = "./Results/Phospho/Figures/TCDS17_Kinases_all.tiff",
  method = "ULM", condition = "t"
  )
```

# Modelling Signalling Network Using PHONEMeS

```{r PHONEMeS, eval=FALSE, include=FALSE}
library(PHONEMeS)
library(visNetwork)
b143.pmes.input <- list()
b143.pmes.input$pps <- b143.DEPhos %>%
  mutate(
    ID = sapply(
      strsplit(rownames(.), "_"), function(x){
        gsub("", "", paste0(x[1], "_", x[2]))
      }
    )
  ) %>%
  filter(ID %in% phonemesPKN$target) %>%
  pull(t, name = ID)
b143.pmes.input$kinS <- b143.Kin$Kinases.list %>%
  mutate(
    Sign = sign(score)
  ) %>%
  pull(Sign, name = source)
b143.pmes.input$kinRm <- b143.Kin$Kinases.list %>%
  filter(abs(score) <= .5) %>%
  pull(source)

CARNIVAL_options <- default_carnival_options(solver = "cplex")
CARNIVAL_options$solverPath <- "~/../../Applications/CPLEX_Studio2211/cplex/bin/arm64_osx/cplex"
CARNIVAL_options$threads <- 8
CARNIVAL_options$timelimit <- 600

b143.pmes.res <- run_phonemes(
  inputObj = b143.pmes.input$kinS,
  measObj = b143.pmes.input$pps,
  rmNodes = b143.pmes.input$kinRm,
  netObj = phonemesPKN,
  carnival_options = CARNIVAL_options
)
b143.pmes.res_pps <- reattach_psites(b143.pmes.res)
b143.pmes.res_prot <- get_protein_network(b143.pmes.res)
b143.pmes.res_mtor <- extract_subnetwork(b143.pmes.res_pps, targets = "MTOR", n_steps = 1)
names(b143.pmes.res_mtor) <- c("Edges", "Nodes")
b143.pmes.res_mtor$Edges <- b143.pmes.res_mtor$Edges %>%
  select(from = Node1, sign = Sign, to = Node2) %>%
  mutate(smooth = T,
         dashes = ifelse(sign > 0, F, T))
b143.pmes.res_mtor$Nodes <- b143.pmes.res_mtor$Nodes %>%
  mutate(
    label = Node,
    type = ifelse(NodeType == "M", "target", "kinase"),
    measured = ifelse(NodeType == "", "interacts", "measured"),
    activity = ifelse(AvgAct > 0, "active", "inactive"),
    color = ifelse(AvgAct > 0, "red", "blue"),
    shape = ifelse(NodeType == "M", "star", "diamond")
  ) %>%
  select(id = Node, label, color, shape, type, measured, activity)
b143.pmes.res_mtor$Net <- visNetwork(
  b143.pmes.res_mtor$Nodes, b143.pmes.res_mtor$Edges,
  main = "MTOR Signalling Network Model", submain = "143B DX-R", width = "100%"
) %>%
  visIgraphLayout(layout = "layout_with_kk", randomSeed = 1) %>%
  visEdges(arrows = "to", color = "dimgrey", width = 2) %>%
#  visNodes(font = "30px arial black bold") %>%
#  visLegend(zoom = F) %>%
  visOptions(highlightNearest = T, nodesIdSelection = T)
b143.pmes.res_mtor$Net
visSave(b143.pmes.res_mtor$Net,
        file = "./Results/Phospho/Figures/b143_MTOR_Sign_n1.html")
webshot2::webshot(
  "./Results/Phospho/Figures/b143_MTOR_Sign_n1.html",
  delay = .2, zoom = 1, vwidth = 992, vheight = 744, quiet = T,
  file = "./Results/Phospho/Figures/b143_MTOR_Sign_n1.png"
)
```


# **Save Results**

```{r Save Results}
KSN <- Kin.net %>%
  mutate(
    sign = sign(mor)
  ) %>%
  select(target, source, sign)
save(
  b143.DEPhos, b143.Kin, b143.Kin_all,
  Saos2.DEPhos, Saos2.Kin, Saos2.Kin_all,
  TCDS17.DEPhos, TCDS17.Kin, TCDS17.Kin_all,
  KSN,
  file = "./Results/Results_Phospho.RData"
)
```
